File: longest-consecutive-sequence.py
# https://leetcode.com/problems/longest-consecutive-sequence/description/
# Save nums to set() & save start num
# Then, start from start
# Time: O(n), Space: O(n)
class Solution:
    def longestConsecutive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        
        used = set()
        start = set()
        
        for n in nums:
            used.add(n)
            if n-1 not in used:
                start.add(n)
            if n+1 in start:
                start.remove(n+1)
        
        longest = 0
        for s in start:
            i = s
            length = 0
            while i in used:
                i += 1
                length += 1
            longest = max(longest, length)
            
        return longest

File: longest-consecutive-sequence.java
// https://leetcode.com/problems/longest-consecutive-sequence/description/
// Add to the HashSet
// Then find seq
// Time: O(n^2) but avg O(n), Space: O(n)
class Solution {
    public int longestConsecutive(int[] nums) {
        if(nums.length==0) return 0;
        
        HashSet<Integer> set = new HashSet<Integer>();
        for(int n : nums) set.add(n);
        
        int ans = 1;
        for(int s : set){
            if(set.contains(s-1)){
                int i = s-1;
                int len = 2;
                while(set.contains(--i)) len++;
                ans = Math.max(ans, len);
            }
        }
        
        return ans;
    }
}
